src/main/java/io/reactivex/internal/operators/flowable/FlowableMergeWithMaybe.java

```diff
@@ -0,0 +1,359 @@
+import org.reactivestreams.*;
+
+import io.reactivex.*;
+import io.reactivex.disposables.Disposable;
+import io.reactivex.internal.disposables.DisposableHelper;
+import io.reactivex.internal.fuseable.SimplePlainQueue;
+import io.reactivex.internal.queue.SpscArrayQueue;
+import io.reactivex.internal.subscriptions.SubscriptionHelper;
+import io.reactivex.internal.util.*;
+import io.reactivex.plugins.RxJavaPlugins;
+
+/**
+ * Merges an Observable and a Maybe by emitting the items of the Observable and the success
+ * value of the Maybe and waiting until both the Observable and Maybe terminate normally.
+ *
+ * @param <T> the element type of the Observable
+ * @since 2.1.10 - experimental
+ */
+public final class FlowableMergeWithMaybe<T> extends AbstractFlowableWithUpstream<T, T> {
+
+    final MaybeSource<? extends T> other;
+
+    public FlowableMergeWithMaybe(Flowable<T> source, MaybeSource<? extends T> other) {
+        super(source);
+        this.other = other;
+    }
+
+    @Override
+    protected void subscribeActual(Subscriber<? super T> observer) {
+        MergeWithObserver<T> parent = new MergeWithObserver<T>(observer);
+        observer.onSubscribe(parent);
+        source.subscribe(parent);
+        other.subscribe(parent.otherObserver);
+    }
+
+    static final class MergeWithObserver<T> extends AtomicInteger
+    implements FlowableSubscriber<T>, Subscription {
+
+        private static final long serialVersionUID = -4592979584110982903L;
+
+        final Subscriber<? super T> actual;
+
+        final AtomicReference<Subscription> mainSubscription;
+
+        final OtherObserver<T> otherObserver;
+
+        final AtomicThrowable error;
+
+        final AtomicLong requested;
+
+        final int prefetch;
+
+        final int limit;
+
+        volatile SimplePlainQueue<T> queue;
+
+        T singleItem;
+
+        volatile boolean cancelled;
+
+        volatile boolean mainDone;
+
+        volatile int otherState;
+
+        long emitted;
+
+        int consumed;
+
+        static final int OTHER_STATE_HAS_VALUE = 1;
+
+        static final int OTHER_STATE_CONSUMED_OR_EMPTY = 2;
+
+        MergeWithObserver(Subscriber<? super T> actual) {
+            this.actual = actual;
+            this.mainSubscription = new AtomicReference<Subscription>();
+            this.otherObserver = new OtherObserver<T>(this);
+            this.error = new AtomicThrowable();
+            this.requested = new AtomicLong();
+            this.prefetch = bufferSize();
+            this.limit = prefetch - (prefetch >> 2);
+        }
+
+        @Override
+        public void onSubscribe(Subscription d) {
+            if (SubscriptionHelper.setOnce(mainSubscription, d)) {
+                d.request(prefetch);
+            }
+        }
+
+        @Override
+        public void onNext(T t) {
+            if (compareAndSet(0, 1)) {
+                long e = emitted;
+                if (requested.get() != e) {
+                    SimplePlainQueue<T> q = queue;
```

Comment:
Are you sure don't need to call `getOrCreateQueue()` here?

Reply:
Yes. If there is no queue yet or is empty, that means T can be emitted immediately and events from the main source won't be reordered. Otherwise the queue exists and the item should be queued.